File: Class-scripts/MATLAB/build_grid3D_implementation.m
%Solving heat equation - D * G * u = fs in 3D
%Spherical source term at the center
%Mohammad Afzal Shadab and Marc Hesse
%Date: 14 March 2022

set_demo_defaults;

%building grid and operators
Grid.xmin = 0; Grid.xmax = 3; Grid.Nx = 50;
Grid.ymin = 0; Grid.ymax = 3; Grid.Ny = 50;
Grid.zmin = 0; Grid.zmax = 3; Grid.Nz = 50;

Grid = build_grid3D(Grid);
[D,G,C,I,M] = build_ops3D(Grid);

%boundary conditions
BC.dof_dir = [Grid.dof_xmin;Grid.dof_xmax];
BC.dof_f_dir= [Grid.dof_f_xmin;Grid.dof_f_xmax];
BC.g = [zeros(length(Grid.dof_xmin),1);zeros(length(Grid.dof_xmax),1)];
BC.dof_neu = [];
BC.dof_f_neu = [];
BC.qb = [];
[B,N,fn] = build_bnd(BC,Grid,I);

L = - D * G;

[X,Y,Z] = meshgrid(Grid.xc,Grid.yc,Grid.zc);
fs = zeros(Grid.N,1);
Xo=(Grid.xmax - Grid.xmin)/2;Yo=(Grid.ymax - Grid.ymin)/2;Zo=(Grid.zmax - Grid.zmin)/2;

fs((X-Xo).^2 + (Y-Yo).^2 + (Z-Zo).^2 < 0.5) = 1.0;  %spherical source term at center

%solving the linear boundary value problem
tic
u = solve_lbvp(L,fs+fn,B,BC.g,N);
toc
%plotting
plotting(Grid,X,Y,Z,u)



function plotting(Grid,X,Y,Z,soln)

%Plotting
%full 3D slice
figure()
xslice = [(Grid.xmax-Grid.xmin)/2];                               % define the cross sections to view
yslice = [(Grid.ymax-Grid.ymin)/2];
zslice = ([(Grid.zmax-Grid.zmin)/2]);

soln = reshape(soln,Grid.Ny,Grid.Nx,Grid.Nz);

slice(X, Y, Z, soln, xslice, yslice, zslice)    % display the slices
xlabel 'x', ylabel 'y', zlabel 'z'
cb = colorbar;                                  % create and label the colorbar
title 'u'
cb.Label.String = 'u ';
%caxis([-120 40])

%1D plots along centerline
figure
title 'Along centerlines'
subplot 311
plot(Grid.xc,soln(Grid.Ny/2,:,Grid.Nz/2),'r','markerfacecolor','w','markersize',6)
xlabel 'x', ylabel 'u'

subplot 312
plot(Grid.yc,soln(:,Grid.Nx/2,Grid.Nz/2),'r','markerfacecolor','w','markersize',6)
xlabel 'y', ylabel 'u'

subplot 313
plot(Grid.zc,reshape(soln(Grid.Ny/2,Grid.Nx/2,:),Grid.Nz,1),'r','markerfacecolor','w','markersize',6)
xlabel 'z', ylabel 'u'

end


%{
disp('Grid.dof_xmin'); Grid.dof_xmin
disp('Grid.dof_xmax');Grid.dof_xmax
disp('Grid.dof_ymin');Grid.dof_ymin
disp('Grid.dof_ymax');Grid.dof_ymax
disp('Grid.dof_zmin');Grid.dof_zmin
disp('Grid.dof_zmax');Grid.dof_zmax
%}

File: Class-scripts/Python/build_grid3D_implementation.py
#Lecture 15 (March 19): 3D Python operator implementation
#Author: Mohammad Afzal Shadab and Marc Hesse

import sys
sys.path.insert(1,'../../HW1-Numerics/Python')
sys.path.insert(1,'../../HW2-BC_LBVP/Python')

#import library and modules
import numpy as np
import matplotlib.pyplot as plt
import pyvista as pv
from pyvista import examples
import time

from classfun import *
from build_gridfun3D import build_grid3D
from build_opsfun3D  import build_ops3D
from solve_lbvpfun_optimized import solve_lbvp
from build_bndfun_optimized import build_bnd
import scipy.sparse as sp

#################################################################################
#Solving the numerical problem
#################################################################################

Grid.xmin = 0; Grid.xmax = 3; Grid.Nx = 50
Grid.ymin = 0; Grid.ymax = 3; Grid.Ny = 50
Grid.zmin = 0; Grid.zmax = 3; Grid.Nz = 50

Grid = build_grid3D(Grid)
[D,G,C,I,M] = build_ops3D(Grid); L = - D*G
fs = np.zeros((Grid.N,1))

#boundary conditions
BC.dof_dir   = np.vstack([Grid.dof_xmin,Grid.dof_xmax])
BC.dof_f_dir = np.vstack([Grid.dof_f_xmin,Grid.dof_f_xmin])
BC.g         = np.vstack([np.ones_like(Grid.dof_xmin),np.zeros_like(Grid.dof_xmax)])

BC.dof_neu   = np.array([])
BC.dof_f_neu = np.array([])
BC.qb        = np.array([])

[B,N,fn] = build_bnd(BC,Grid,I)

[X,Z,Y] = np.meshgrid(Grid.xc,Grid.zc,Grid.yc)
X_col = np.transpose([X.flatten()]);Y_col = np.transpose([Y.flatten()]);Z_col = np.transpose([Z.flatten()])

#source term
fs = np.zeros((Grid.N,1))
Xo = (Grid.xmax-Grid.xmin)/2; Yo=(Grid.ymax-Grid.ymin)/2; Zo=(Grid.zmax-Grid.zmin)/2
fs[((X_col-Xo)**2+(Y_col-Yo)**2+(Z_col-Zo)**2)<0.5] = 1.0 

t1 = time.perf_counter()
u = solve_lbvp(L, fs+fn, B, BC.g, N)
t2 = time.perf_counter()
print('Time elapsed',t2-t1)
U = u.reshape(Grid.Nz,Grid.Nx,Grid.Ny)


#####################################################################################
'''
Plotting
'''

# Create and plot structured grid
grid = pv.StructuredGrid(X,Y,Z)

# Add the data values to the cell data
grid.point_data["values"] = U.flatten(order="F")  # Flatten the array!

#grid.plot(show_edges=True,show_grid=True)

slices = grid.slice_orthogonal(x=1.5, y=1.5, z=1.5)
slices.plot(show_edges=True,show_grid=True)
plotter = pv.Plotter(off_screen=True)

plotter.add_mesh(slices, show_edges=True)
plotter.add_axes(line_width=5, labels_off=False)
plotter.show_grid()
#plotter.show(screenshot='heat_equation.png')
plotter.save_graphic("heat_equation.pdf") 

#####################################################################################
#plotting 
fig, (ax1, ax2, ax3) = plt.subplots(3,1, figsize=(10,7))
fig.suptitle(f'Numerical results on centerlines')

ax1.plot(Grid.xc,U[int(Grid.Ny/2),:,int(Grid.Nz/2)],'r-',label=r'$Numerical$')
ax1.set_ylabel('$d2f$')
ax1.set_xlabel('$x$')

ax2.plot(Grid.xc,U[:,int(Grid.Nx/2),int(Grid.Nz/2)],'r-',label=r'$Numerical$')
ax2.set_ylabel('$d2f$')
ax2.set_xlabel('$y$')

ax3.plot(Grid.xc,U[int(Grid.Ny/2),int(Grid.Nx/2),:],'r-',label=r'$Numerical$')
ax3.set_ylabel('$d2f$')
ax3.set_xlabel('$z$')


plt.tight_layout()
